Back translation helper updates

bobeaton opened the pull request:
fixes to various issues after using the Back Translation helper on a real project -- mostly making it more user-friendly and dealing with issues about what happens if the user clicks in different verses in a project in Paratext while the BTHelper dialog/plugin is active (and monitoring the "current verse"). Also dealing w/ differences between the source and target projects -- markers used and such)

Commit:
updated SpellFixer to support RtL languages (now that it's used natively in the Ptx plugin for BTs)

Commit:
when querying for the current target text, we can have a failure if the user entered a range of verses by using the up/down control at the top of Ptx window (bkz it thinks the verse is just the one specified by the up/down control, whereas the collection of USFMTokens is based on the range... so triangulate to find the range from just the single verse if that's what we have

Commit:
fix several bugs related to the dialog thrashing if the user clicks off the dialog and changes things in Ptx, or overwriting data unintentionally

Commit:
added a button to be able to 'pin' the form as topmost (so it'll still show above Ptx if selected)

Commit:
don't preview online translators or compound or primary-fallback converters (it takes too much time)

Commit:
moved the 'stay on top' pin to the underlying control

Commit:
change WriteToTarget so that it returns a bool so the ctrl can stop moving forward if it returns false.

Commit:
missed a file that had changes for moving the pin to top to the control

Commit:
added some debugging statements so I can figure out why we're overwriting data after editing manually in Ptx.

Commit:
figured out the fix for overwriting data: don't mark the data dirty until the form comes back in focus. (if you do it when it loses focus, it may get marked clean thru the call from Ptx on change of verse (if the user is editing in different verses in Ptx). By waiting until we get back, we can guarantee it gets refreshed when the Form has focus (by which Ptx editing is finished)

Commit:
the repositioning of the form after startup (from where it was when we shut down last time) wasn't working. Changed it to wait for the Load event before setting the boundaries (rather than the ctor)

Commit:
update LangVersion to 'latest' in projects that are behind

Commit:
don't bother asking the user if they want to requery the data; just do it for them (it shouldn't overwrite any edited text)

Commit:
updated projects to the new EcCore nuget 0.5.5

Commit:
update to latest ptx plugin nuget

Commit:
adjusted how we're marking text dirty -- via the ScriptureDataChangedHandler methods rather than just that we've 'Deactivated' (with the former, we *know* the text has changed).

Commit:
clean up several things the compiler is warning about (unused functions or standard naming for variables)

Commit:
redo CalculateTargetTokens so that it always replicates the source markers to replace the target markers. We might not have enough translated lines for the source, but in any case, the source is what we're trying to back-translate; not merging with what's currently in the target (which could be problematic).

Commit:
re-created a number of Forms used in the 'SpellFixer' (aka. Find-Replace) project to make them more current and behave nicer (also got rid of supporting legacy encodings -- anyone not using Unicode by now needs to feel the pain). Also includes adding a dummy rule (\u001f => \u001f) to the underlying cc table so that it won't generate the 'there are no rules' exception.

Commit:
if the Paratext BT Helper plugin is installed, install SpellFixer (aka. FindReplace, which is now implicitly available in the plugin via the Add Substitution button and related context menu) and SetPath, so the cc engine (used by SpellFixer) is reachable

Commit:
add scaffolding for RegisterForNotification with the BTHelperControl so we can get updates to the TargetTranslation text box (and warn user about discrepancies in the number of lines translated)

Commit:
updated the writing to the textBoxStatus (it needed Application.DoEvents()) for it to show

Commit:
fixed a number of bugs related working with a target project that has no existing data and discrepancies between versification between source and target (source is combined, target are separate, and vice-versa)

Commit:
added a text box to the Ptx plugin Form so that we could display status (e.g. differences between the number of translated lines between source and target and warnings about staying on verses (if the user clicks on another one in Ptx) if the text was modified (lest we lose changes unintentionally)

Commit:
fix the case where Paratext doesn‘t consider the verse number in a \va...\va* inline segment as publishable text (in need of being translated, which it is) if it comes immediately after a \v marker...

Commit:
prevent an exception if the source has no data (not expected, but can happen -- e.g. if the person clicks Next to a verse that hasn't been translated in the source yet).

Commit:
the same fix I did for reading source tokens and supporting the aberent case of a \va immediately after a \v not behaving the same way needed to be fixed in several other places also (the reading of target tokens and in the calculating of the data to write to the target project during 'Next' and 'Save Changes')

Commit:
updated to 0.5.6 of the EncConverters core

Commit:
make the status line better English when singular vs. plural

Commit:
updated installer version to 5.1.2

Commit:
fix error: "error CS8652: The feature 'constant interpolated strings' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version."
(it never needed to be a 'constant interpolated string' anyway.)

Commit:
update license text to the correct year

Commit:
add a new intermediate class that can create different TranslationHelperForm instances based on the font + converter, so that we can support multiple different ones at the same time

Commit:
keep track of location so it doesn't shift positions for each translation

Commit:
a bunch of changes suggested by VS2022 for newer version of C#

Commit:
moved the 'reset' ribbon button to the end, so it's more clear

Commit:
added support to pub and the main Word app for the new intermediate class that allows for multiple TranslationHelperForm instances based on Font + Converter

Commit:
update the version reference of SpellFixer in its merge module, which is launched from reflection/COM and needs to match

Commit:
backTranslationHelperCtrl changed behavior to send the button pressed *after* writing the text out, so now 'MoveToNext' also means that we want to return 'ReplaceOnce' to the caller

Commit:
include the EcTranslators and SpellFixerEc merge modules if any of the SILConverters for Office features are being installed (since those are necessary for the TranslationHelperForm and the Find-Replace features in it

Commit:
change nuget for EcCore to 0.5.7

Commit:
save the current window location and size when closing the form -- for the Word plug-in

Commit:
added a TranslationHelperForm to the Xml Document converter (so it could be used, e.g. w/ Transcelerator adaptations.

Commit:
fix the command line option for XML converter so that it can still do the entire thing w/o the translation helper dialog (some may find it easier to just let Bing/DeepL do the entire file and then clean it up afterwards)

Commit:
fixed the logic in the XML converter so that clicking 'Skip' means it won't convert the entry. Also added the *.config files to the merge module for SilConvertersXml.exe so on install, it'll redirect the System.Runtime.CompilerServices.Unsafe dll to v7, which is installed

Commit:
removing RegisterForComInterop for Debug x64 build hoping to fix the GHA build (though, it isn't really used for anything, since we have to sign it using TeamCity)

Commit:
removed some unnecessary (and breaking) "Properties" dereferences

Commit:
added another way to get font listing which isn't as verbose as 'on entire machine' and also changed it so that if the user doesn't have anything selected, then functions default to 'whole document'

Commit:
IcuRegex doesn't work for certain regex escape codes (e.g. \u200c or \u200d -- the zero-width (non-) joiners), so I'm giving up on using it for the regex searching in Word functionality and replacing it with .Net's version of regex, which is functioning nicely...

Commit:
updated to new nuget 0.5.8

Commit:
adjusted the AddNewProject form in SpellFixer so that it resizes properly

Commit:
fixed issue with setting progress bar during background working (being on the wrong thread)

Commit:
changes (mostly to app.config) for redirecting DLLs for the new Google Translate EncConverter

Commit:
updated to new 0.6.0 version of EncConverters core (w/ google translate plugin)

papeh commented on Installer/SetupSILConverters/SetupSILConverters.wixproj:
```diff
@@ -17,10 +17,10 @@
     <OutputPath>bin\$(Platform)\$(Configuration)\</OutputPath>
     <IntermediateOutputPath>obj\$(Platform)\$(Configuration)\</IntermediateOutputPath>
-    <DefineConstants>ApplicationName=SILConverters;SafeApplicationName=SILConverters;TruncatedVersion=5.1;VersionNumber=5.1.1.0;UpgradeCode=8A50226C-84D0-4CAF-AF8F-B0284643C415;Year=2022;Manufacturer=SIL International;EcDistFilesPath=$(EcDistFilesPath)</DefineConstants>
+    <DefineConstants>ApplicationName=SILConverters;SafeApplicationName=SILConverters;TruncatedVersion=5.1;VersionNumber=5.1.2.0;UpgradeCode=8A50226C-84D0-4CAF-AF8F-B0284643C415;Year=2022;Manufacturer=SIL International;EcDistFilesPath=$(EcDistFilesPath)</DefineConstants>
```
`Year=2022` There has to be a way to centralize some of these constants.

bobeaton replied:
if you have a suggestion, I'm open. I found out about having a "Directory.Build.props" file in the solution root and I do have that (read: I have to change it in both places). But for some reason, they aren't visible in wix projects where these are needed.

papeh replied:
Hmm. In [genericinstaller](https://github.com/sillsdev/genericinstaller/blob/master/BaseInstallerBuild/buildBaseInstaller.bat), we pass this on the command line, but your setup is probably different.

I've tried moving `<DefineConstants>` out of the `<PropertyGroup Condition=Release|x86>`, but if you edit that in VisualStudio, you'll probably be right back where you started.

Did you try putting `<DefineConstants>` in Directory.Build.props, or only `<Year>2023</Year>`? But if `<DefineConstants>Year=$(Year)</DC>` didn't work, I'm out of ideas.

bobeaton replied:
yeah... it just doesn't want to go from Directory.Build.props to a wix project w/ wix 3.11.2... there is a v4 of wix, but I couldn't upgrade, bkz "Package 'wix 4.0.0' has a package type 'DotnetTool' that is not supported by project 'Installer\SEC Setup'" (which I think is looking for .net core functionality which this code base doesn't support). Anyway, these defines are defined in the VS UI and stored in the proj file, but as long as I remember to change them in both places (x86 and x64, which have different upgrade codes anyway), then it works okay. 

maybe a project for another release (when I have more time to research it)

papeh commented on src/ParatextPluginBackTranslationHelper/Properties/Resources.resx:
```diff
@@ -118,7 +118,13 @@
+  </data>
   <data name="SplashScreenImage" type="System.Resources.ResXFileRef, System.Windows.Forms">
-    <value>..\Resources\Image1.bmp;System.Drawing.Bitmap, System.Drawing, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a</value>
+    <value>..\Resources\SplashScreenImage.bmp;System.Drawing.Bitmap, System.Drawing, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a</value>
```
Hooray! A descriptive name! 🎉

papeh commented on src/SILConvertersWordML/TestBwdc/UnitTest_PtxBackTrHelper.cs:
```diff
@@ -17,22 +17,26 @@ namespace TestBwdc
         [TestCase("SingleVerse", "SingleVerseMissingInTarget")]
+        /* these are all now failing and I don't think they matter anymore. We aren't now completely 
+         * overwriting the target text with the translated text, rather than merging with what's
+         * in the target presently.
```
> We aren't now . . . rather than

This is unclear to me. Could it be reworded? E.g., do you mean, We are no longer completely overwriting, but rather we are merging?

bobeaton replied:
Updated as follows:
```
        /* the test cases below are all now failing, but they don't matter anymore. They were for 
         * when we were trying to merge the markers from the source with any possible ones existing 
         * in the target. This was a fool's errand (too many possible differences). 
         * Now we just completely overwrite the text in the target project with the same markers from
         * the source project (with the source text replaced by the translated text). So what these 
         * tests were trying to validate doesn't happen anymore.
```

papeh commented on src/SilConvertersXML/Properties/Resources.resx:
```diff
@@ -216,7 +216,12 @@ to the left. Right-click on an item in the tree to select it.
+(e.g. if the list is large and you didn't finish it last time), 
+you can select the item you want to start with and when you click 
+the 'Convert and Save XML Document' button, you'll be asked if 
+you want to start the conversation from that item or not.</value>
```
> conversation

conversion?

bobeaton replied:
haha... good catch!
fixed

papeh reviewed (commented):
Looked at most paracode. Will look at .cs files next.

papeh commented on src/ParatextPluginBackTranslationHelper/BackTranslationHelperForm.cs:
```diff
@@ -222,6 +318,51 @@ private string CurrentSourceData
+        // So... if IsPubliableVernacular is false, at least check if this is that case, and return true, so we'll try to 
+        // translate it as the others are (bkz we only send IsPub text segments for translation)
+        private static bool IsPublishableVernacular(IUSFMTextToken t, List<IUSFMToken> tokens)
+        {
```
a unit test for this would be helpful

bobeaton replied:
it's the "CombinedInSourceAndTargetWithVaOverride" TestCase in the It_Can_Determine_Correct_Update_To_Target_Project_For_Other_Cases test

papeh commented on src/ParatextPluginBackTranslationHelper/BackTranslationHelperForm.cs:
```diff
@@ -338,30 +505,53 @@ void IBackTranslationHelperDataSource.MoveToNext()
-                    return; // nothing to do
+                    // if we don't already have it... it's probably because something was changed in the project
+                    //  external to this form (we lose the data if we get activated).
+                    System.Diagnostics.Debug.WriteLine($"PtxBTH: RequeryWarning");
```
did you mean to include _verseReference in this $"{string}"?

bobeaton replied:
added both (the single version based on the up-down control) and the range if the actual verses are combined:
                    System.Diagnostics.Debug.WriteLine($"PtxBTH: RequeryWarning: _verseReference: {_verseReference}, _versesReference: {_versesReference}");

papeh reviewed (commented):
Overall, looks good so far.

Commit:
fixed https://github.com/silnrsi/silconverters/issues/8 (labels on Bulk Word Doc converter and several other features)

Commit:
fixed https://github.com/silnrsi/silconverters/pull/13#discussion_r1154875690

Commit:
fixed https://github.com/silnrsi/silconverters/pull/13#discussion_r1154879092

Commit:
In Spellfixer/AddProject dialog, the validity checks I was doing in ButtonOk_Click didn't stop the dialog from closing... moved them to AddNewProjectForm_FormClosing, where the closing can be canceled.
Also added proper tab order.
Also, keep the 'Add Project' button all the time; not just if there isn't one.

Commit:
added some notes on how to build and upgrade the EcCore package

Commit:
fixed https://github.com/silnrsi/silconverters/pull/13#discussion_r1157398881

Commit:
attempted to fix this: https://github.com/silnrsi/silconverters/pull/13#discussion_r1157405726
but was unsuccessful except to fix the wrong year

Commit:
center the spellfixer dialog boxes when they appear (in case the translation dialog is on the edge (more likely) and pinned to the top (which would leave these hidden behind it)

Commit:
oops. accidentally undid the fix recommended for: https://github.com/silnrsi/silconverters/issues/7

Commit:
ignoring some files that shouldn't have been checked in earlier

Commit:
make sure the word.doc and .app objects are released so we don't leave around a stranded winword.exe process

Commit:
if the user changes some of the 'advanced' options, then we need to reload the documents

Commit:
Change how the 'combineAllRunsIntoASingleRun' works so that if the first run(s) contains only punctuation, they don't define the font to use for the full run (since they are often in a different font in Word)

Commit:
if the user chooses a Translator EncConverter, suggest that they use 'combineIsoFormattedParagraphs'

Commit:
added a unit test to make sure that if we're converting paragraphs within table cells, that they are done properly.

Commit:
also don't want to use whitespace or digits as the indication of what font to use for a paragraph

Commit:
lang version wasn't set for x86 build
